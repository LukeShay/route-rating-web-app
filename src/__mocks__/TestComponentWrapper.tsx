/* eslint @typescript-eslint/no-empty-function: 0, @typescript-eslint/explicit-function-return-type: 0 */
import React from "react";
import { IGymsContextState } from "../context/gyms/gymsStore";
import { IViewContextState } from "../context/view/viewStore";
import { IUserContextState } from "../context/user/userStore";
import { StaticRouter } from "react-router";
import GymsStoreMock from "./gymsStoreMock";
import ViewStoreMock from "./viewStoreMock";
import UserStoreMock from "./userStoreMock";

export interface ITestComponentWrapperProps {
  children: React.ReactNode;
  gymsState: IGymsContextState;
  location: string;
  userState: IUserContextState;
  viewState: IViewContextState;
}

const TestComponentWrapper: React.FC<ITestComponentWrapperProps> = ({
  children,
  gymsState,
  location,
  userState,
  viewState
}): JSX.Element => {
  return (
    <StaticRouter location={`https://lukeshay.com/${location}`}>
      <GymsStoreMock state={gymsState} dispatch={() => {}}>
        <ViewStoreMock state={viewState} dispatch={() => {}}>
          <UserStoreMock state={userState} dispatch={() => {}}>
            {children}
          </UserStoreMock>
        </ViewStoreMock>
      </GymsStoreMock>
    </StaticRouter>
  );
};

export default TestComponentWrapper;
