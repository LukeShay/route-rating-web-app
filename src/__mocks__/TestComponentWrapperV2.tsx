/* eslint @typescript-eslint/no-empty-function: 0, @typescript-eslint/explicit-function-return-type: 0 */
import React from "react";
import { IGymsContextState } from "../context/gyms/gymsStore";
import { IUserContextState } from "../context/user/userStore";
import { IViewContextState } from "../context/view/viewStore";
import GymsStoreMock from "./gymsStoreMock";
import ViewStoreMock from "./viewStoreMock";
import UserStoreMock from "./userStoreMock";

export interface ITestComponentWrapperV2Props {
  children: React.ReactNode;
  gymsState: IGymsContextState;
  userState: IUserContextState;
  viewState: IViewContextState;
}

const TestComponentWrapperV2: React.FC<ITestComponentWrapperV2Props> = ({
  children,
  gymsState,
  userState,
  viewState
}): JSX.Element => {
  return (
    <GymsStoreMock state={gymsState} dispatch={() => {}}>
      <ViewStoreMock state={viewState} dispatch={() => {}}>
        <UserStoreMock state={userState} dispatch={() => {}}>
          {children}
        </UserStoreMock>
      </ViewStoreMock>
    </GymsStoreMock>
  );
};

export default TestComponentWrapperV2;
