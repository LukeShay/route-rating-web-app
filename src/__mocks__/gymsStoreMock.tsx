import { GymsContext, IGymsContextState } from '../context/gyms/gymsStore';
import React from 'react';

export interface IGymsStoreMockProps {
  state: IGymsContextState;
  dispatch: any;
}

const GymsStoreMock: React.FC<IGymsStoreMockProps> = ({
  state,
  dispatch,
  children,
}): JSX.Element => (
  <GymsContext.Provider value={{ state, dispatch }}>
    {children}
  </GymsContext.Provider>
);

export default GymsStoreMock;
