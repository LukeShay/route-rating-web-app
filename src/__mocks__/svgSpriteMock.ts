export default {
  id: 'svg-sprite-id',
  viewBox: '0 0 1 1',
  url: '/sprite.svg#svg-sprite-id',
};
