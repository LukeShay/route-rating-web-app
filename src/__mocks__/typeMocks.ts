import { Gym, GymPageType, Route, User, Wall, WallTypes } from '../types';
import { GymsContextState } from '../context/gyms/gymsStore';
import { ViewContextState } from '../context/view/viewStore';
import { UserContextState } from '../context/user/userStore';

export const testRouteOne: Route = {
  averageGrade: 'GRADE_5_9',
  averageRating: 4,
  gymId: 'gym1',
  holdColor: 'Blue',
  id: 'route1',
  name: 'RouteOne',
  setter: 'LukeOne',
  types: [WallTypes.TOP_ROPE, WallTypes.LEAD],
  wallId: 'wall1',
};

export const testRouteTwo: Route = {
  averageGrade: 'GRADE_5_12ab',
  averageRating: 4,
  gymId: 'gym1',
  holdColor: 'Red',
  id: 'route2',
  name: 'RouteTwo',
  setter: 'LukeTwo',
  types: [WallTypes.TOP_ROPE, WallTypes.LEAD],
  wallId: 'wall1',
};

export const testWallOne: Wall = {
  gymId: 'gym1',
  id: 'wall1',
  name: 'Wall1',
  routes: [testRouteOne, testRouteTwo],
  types: [WallTypes.LEAD, WallTypes.TOP_ROPE],
};

export const testWallTwo: Wall = {
  gymId: 'gym1',
  id: 'wall2',
  name: 'Wall2',
  routes: [],
  types: [WallTypes.LEAD, WallTypes.TOP_ROPE],
};

export const testGymOne: Gym = {
  address: '1234 Five Street',
  authorizedEditors: ['id'],
  city: 'Ames',
  email: 'abc@d.com',
  id: 'gym1',
  logoUrl: '',
  name: 'Gym One',
  phoneNumber: '1234567890',
  photoUrl: '',
  state: 'Iowa',
  walls: [testWallOne, testWallTwo],
  website: 'lukeshay.com',
  zipCode: '50014',
};

export const testGymTwo: Gym = {
  address: '1234 Five Street',
  authorizedEditors: ['id'],
  city: 'Ames',
  email: 'abc@d.com',
  id: 'gym2',
  logoUrl: '',
  name: 'Gym Two',
  phoneNumber: '1234567890',
  photoUrl: '',
  state: 'Iowa',
  walls: [testWallOne, testWallTwo],
  website: 'lukeshay.com',
  zipCode: '50014',
};

export const testUserEditor: User = {
  authority: 'ADMIN',
  city: 'Ames',
  country: 'USA',
  email: 'email',
  firstName: 'Name',
  id: 'id',
  lastName: 'Last',
  password: 'password',
  phoneNumber: '1111111111',
  role: 'ADMIN_ROLE',
  session: null,
  state: 'IA',
  username: 'username',
};

export const mockGymsState: GymsContextState = {
  page: { content: [testGymOne, testGymTwo] } as GymPageType,
};

export const darkDesktopState: ViewContextState = {
  mobile: false,
  theme: 'DARK_THEME',
};

export const signedInEditorState: UserContextState = {
  user: testUserEditor,
};

export const signedOutState: UserContextState = {
  user: null,
};
