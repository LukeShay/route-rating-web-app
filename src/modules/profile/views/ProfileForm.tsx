import AccountCircleIcon from '@material-ui/icons/AccountCircle';
import React from 'react';
import * as UserActions from '../../../context/user/userActions';
import { useUserContext } from '../../../context/user/userStore';
import { ButtonEvent, InputEvent, User } from '../../../types';
import Button from '../../common/buttons/ButtonSecondary';
import Form from '../../common/forms/Form';
import Input from '../../common/inputs/Input';
import * as ResponseUtils from '../../../utils/responseUtils';
import { ErrorResponse } from '../../../types/responses';
import { toast } from 'react-toastify';
import * as RegexUtils from '../../../utils/regexUtils';

export interface IPropsProfileForm {
  user: User;
}

const ProfileForm: React.FC<IPropsProfileForm> = ({ user }): JSX.Element => {
  const { dispatch: userDispatch } = useUserContext();
  const [firstName, setFirstName] = React.useState<string>(user.firstName);
  const [lastName, setLastName] = React.useState<string>(user.lastName);
  const [email, setEmail] = React.useState<string>(user.email);
  const [emailMessage, setEmailMessage] = React.useState<string>('');
  const [city, setCity] = React.useState<string>(user.city);
  const [cityMessage] = React.useState<string>('');
  const [state, setState] = React.useState<string>(user.state);
  const [stateMessage, setStateMessage] = React.useState<string>('');
  const [username, setUsername] = React.useState<string>(user.username);
  const [usernameMessage, setUsernameMessage] = React.useState<string>('');
  const [phoneNumber, setPhoneNumber] = React.useState<string>(
    user.phoneNumber
  );
  const [phoneNumberMessage, setPhoneNumberMessage] = React.useState<string>(
    ''
  );
  const [password, setPassword] = React.useState<string>('');
  const [passwordMessage, setPasswordMessage] = React.useState<string>('');
  const [repeatPassword, setRepeatPassword] = React.useState<string>('');
  const [repeatPasswordMessage, setRepeatPasswordMessage] = React.useState<
    string
  >('');
  const [loading, setLoading] = React.useState<boolean>(false);

  const validatePassword = (): boolean => {
    if (password.length === 0) {
      setPasswordMessage('');
      return false;
    } else if (password.length < 8) {
      setPasswordMessage('Password must be at least 8 characters long.');
      return false;
    } else if (!RegexUtils.containsLowercase(password)) {
      setPasswordMessage('Password must contain a lower case letter.');
      return false;
    } else if (!RegexUtils.containsUppercase(password)) {
      setPasswordMessage('Password must contain an upper case letter.');
      return false;
    } else if (!RegexUtils.containsNumber(password)) {
      setPasswordMessage('Password must contain a number.');
      return false;
    } else if (!RegexUtils.containsSpecialCharacter(password)) {
      setPasswordMessage('Password must contain a special character.');
      return false;
    } else {
      setPasswordMessage('');
      return true;
    }
  };

  const validateEmail = (): boolean => {
    if (email.length === 0) {
      setEmailMessage('');
      return false;
    } else if (!RegexUtils.validEmail(email)) {
      setEmailMessage('Invalid email.');
      return false;
    } else {
      setEmailMessage('');
      return true;
    }
  };

  const validatePhoneNumber = (): boolean => {
    if (phoneNumber.length === 0) {
      setPhoneNumberMessage('');
      return false;
    } else if (
      !RegexUtils.containsOnlyNumbers(phoneNumber) ||
      phoneNumber.length > 10 ||
      phoneNumber.length < 10
    ) {
      setPhoneNumberMessage('Invalid phone number. Format: ##########');
      return false;
    } else {
      setPhoneNumberMessage('');
      return true;
    }
  };

  React.useEffect(() => {
    validatePassword();
  }, [password]);

  React.useEffect(() => {
    validateEmail();
  }, [email]);

  React.useEffect(() => {
    validatePhoneNumber();
  }, [phoneNumber]);

  React.useEffect(() => {
    if (repeatPassword !== password) {
      setRepeatPasswordMessage('Passwords do not match.');
    } else {
      setRepeatPasswordMessage('');
    }
  }, [password, repeatPassword]);

  async function handleSignOutClick(): Promise<void> {
    await UserActions.signOut(userDispatch);
  }

  async function handleSubmit(event: ButtonEvent): Promise<void> {
    event.preventDefault();
    setLoading(true);

    if (validatePhoneNumber() && validateEmail()) {
      const response = await UserActions.updateUser(userDispatch, {
        city,
        email: email.toLowerCase(),
        firstName,
        lastName,
        password,
        phoneNumber,
        state,
        username,
      } as User);

      if (!response) {
        ResponseUtils.toastIfNotOk(
          response,
          'Error updating user. Please try again.'
        );
      } else if (!ResponseUtils.isOk(response)) {
        const body: User = await response.json();

        if (body.email) {
          setEmailMessage(body.email);
        }
        if (body.username) {
          setUsernameMessage(body.username);
        }
        if (body.phoneNumber) {
          setPhoneNumberMessage(body.phoneNumber);
        }
        if (body.password) {
          setPasswordMessage(body.password);
        }
        if (body.state) {
          setStateMessage(body.state);
        }
      } else {
        toast.success('User updated.');
      }
    }
    setLoading(false);
  }

  const handleChange = async (event: InputEvent): Promise<void> => {
    event.preventDefault();
    const { id, value } = event.target;

    if (id === 'firstName') {
      setFirstName(value);
    } else if (id === 'lastName') {
      setLastName(value);
    } else if (id === 'email') {
      setEmail(value);
    } else if (id === 'username') {
      setUsername(value);
    } else if (id === 'phoneNumber') {
      setPhoneNumber(value);
    } else if (id === 'city') {
      setCity(value);
    } else if (id === 'state') {
      setState(value);
    } else if (id === 'password') {
      setPassword(value);
    } else if (id === 'repeatPassword') {
      setRepeatPassword(value);
    }
  };

  const formInputs: JSX.Element = (
    <React.Fragment>
      <Input
        autoCapitalize="true"
        autoComplete="first-name"
        id="firstName"
        onChange={handleChange}
        placeholder="First Name"
        type="text"
        value={firstName}
      />
      <Input
        autoCapitalize="true"
        autoComplete="last-name"
        id="lastName"
        onChange={handleChange}
        placeholder="Last Name"
        type="text"
        value={lastName}
      />
      <Input
        autoComplete="email"
        helpText={emailMessage}
        id="email"
        onChange={handleChange}
        placeholder="Email"
        type="text"
        value={email}
      />
      <Input
        autoComplete="username"
        helpText={usernameMessage}
        id="username"
        onChange={handleChange}
        placeholder="Username"
        type="text"
        value={username}
      />
      <Input
        autoComplete="phone-number"
        helpText={phoneNumberMessage}
        id="phoneNumber"
        onChange={handleChange}
        placeholder="Phone Number"
        type="text"
        value={phoneNumber}
      />
      <Input
        autoComplete="city"
        helpText={cityMessage}
        id="city"
        onChange={handleChange}
        placeholder="City"
        type="text"
        value={city}
      />
      <Input
        autoComplete="state"
        helpText={stateMessage}
        id="state"
        onChange={handleChange}
        placeholder="State"
        type="text"
        value={state}
      />
      <Input
        placeholder="Password"
        id="password"
        value={password}
        onChange={handleChange}
        helpText={passwordMessage}
        type="password"
        autoComplete="password"
      />
      <Input
        placeholder="Repeat Password"
        id="repeatPassword"
        value={repeatPassword}
        onChange={handleChange}
        helpText={repeatPasswordMessage}
        type="password"
      />
    </React.Fragment>
  );

  const title: JSX.Element = (
    <div style={{ display: 'inline' }}>
      <div style={{ float: 'left', marginRight: '25px', marginTop: '5px' }}>
        Your profile
      </div>
      <div style={{ float: 'right', marginLeft: '25px' }}>
        <Button
          id="signOut"
          onClick={handleSignOutClick}
          type="button"
          variant="outlined"
        >
          Sign out
        </Button>
      </div>
    </div>
  );

  return (
    <Form
      id="profile"
      buttonText="Update Account"
      formInputs={formInputs}
      handleSubmit={handleSubmit}
      icon={<AccountCircleIcon />}
      title={title}
      disabled={loading}
    />
  );
};

export default ProfileForm;
