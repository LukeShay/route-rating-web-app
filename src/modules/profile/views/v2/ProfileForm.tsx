import React from 'react';
import UserForm from './UserForm';
import { User } from '../../../../types';
import * as UserActions from '../../../../context/user/userActions';
import * as ResponseUtils from '../../../../utils/responseUtils';
import { toast } from 'react-toastify';
import { useUserContext } from '../../../../context/user/userStore';
import * as UserUtils from '../../../../utils/userUtils';
import { RecaptchaUser } from '../../../../types';

export interface IProfileFormProps {
  user: User;
}

const ProfileForm: React.FC<IProfileFormProps> = ({ user }): JSX.Element => {
  const { dispatch: userDispatch } = useUserContext();
  const [currentUser, setCurrentUser] = React.useState<RecaptchaUser>({
    recaptchaResponse: '',
    ...user,
  } as RecaptchaUser);

  const [helpMessages, setHelpMessages] = React.useState<RecaptchaUser>({
    recaptchaResponse: '',
    ...UserUtils.createEmptyUser(),
  } as RecaptchaUser);

  async function handleSubmit(updatedUser: RecaptchaUser): Promise<void> {
    setCurrentUser(updatedUser);
    const response = await UserActions.updateUser(userDispatch, updatedUser);

    if (
      !ResponseUtils.toastIfNotOk(
        response,
        'Error updating user. Please try again.'
      )
    ) {
      if (response instanceof Response) {
        const errors: RecaptchaUser = await response.json();

        setHelpMessages(errors);
      } else {
        toast.error('Error updating user. Please try again.');
      }
    } else {
      toast.success('User updated.');
    }
  }

  async function handleSignOutClick(): Promise<void> {
    await UserActions.signOut(userDispatch);
  }

  return (
    <UserForm
      buttonText="Sign out"
      headerText="Your profile"
      help={helpMessages}
      id="profile"
      onButtonClick={handleSignOutClick}
      onSubmit={handleSubmit}
      recaptcha={false}
      submitButtonText="Update profile"
      user={currentUser}
    />
  );
};

export default ProfileForm;
