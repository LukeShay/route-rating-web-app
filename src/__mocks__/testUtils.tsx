import { render, RenderResult } from '@testing-library/react';
import React from 'react';
import TestComponentWrapperV2 from './TestComponentWrapperV2';
import * as TypeMocks from './typeMocks';

interface WrapperProps {
  children: React.ReactNode;
}

const SignedInDesktop: React.FC<WrapperProps> = ({ children }): JSX.Element => {
  return (
    <TestComponentWrapperV2
      gymsState={TypeMocks.mockGymsState}
      userState={TypeMocks.signedInEditorState}
      viewState={TypeMocks.darkDesktopState}
    >
      {children}
    </TestComponentWrapperV2>
  );
};

export const renderSignedInDesktop = (ui, options): RenderResult =>
  render(ui, { wrapper: SignedInDesktop, ...options });

const SignedOutDesktop: React.FC<WrapperProps> = ({
  children,
}): JSX.Element => {
  return (
    <TestComponentWrapperV2
      gymsState={TypeMocks.mockGymsState}
      userState={TypeMocks.signedOutState}
      viewState={TypeMocks.darkDesktopState}
    >
      {children}
    </TestComponentWrapperV2>
  );
};

export const renderSignedOutDesktop = (ui, options): RenderResult =>
  render(ui, { wrapper: SignedOutDesktop, ...options });

export * from '@testing-library/react';
