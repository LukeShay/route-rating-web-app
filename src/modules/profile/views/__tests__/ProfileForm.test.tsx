import * as TypeMocks from '../../../../__mocks__/typeMocks';
import ProfileForm from '../ProfileForm';
import React from 'react';
import { shallow } from '../../../../../configs/setupEnzyme';
import { shallowToJson } from 'enzyme-to-json';

describe('ProfileForm', () => {
  it('should render correctly.', async () => {
    const wrapper = shallow(<ProfileForm user={TypeMocks.testUserEditor} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
