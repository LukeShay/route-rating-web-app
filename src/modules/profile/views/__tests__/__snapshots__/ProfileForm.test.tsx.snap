// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileForm should render correctly. 1`] = `
<Form
  buttonText="Update Account"
  disabled={false}
  formInputs={
    <React.Fragment>
      <Input
        autoCapitalize="true"
        autoComplete="first-name"
        error={true}
        fullWidth={true}
        helpText=""
        id="firstName"
        name=""
        onChange={[Function]}
        placeholder="First Name"
        type="text"
        value="Name"
      />
      <Input
        autoCapitalize="true"
        autoComplete="last-name"
        error={true}
        fullWidth={true}
        helpText=""
        id="lastName"
        name=""
        onChange={[Function]}
        placeholder="Last Name"
        type="text"
        value="Last"
      />
      <Input
        autoComplete="email"
        error={true}
        fullWidth={true}
        helpText=""
        id="email"
        name=""
        onChange={[Function]}
        placeholder="Email"
        type="text"
        value="email"
      />
      <Input
        autoComplete="username"
        error={true}
        fullWidth={true}
        helpText=""
        id="username"
        name=""
        onChange={[Function]}
        placeholder="Username"
        type="text"
        value="username"
      />
      <Input
        autoComplete="phone-number"
        error={true}
        fullWidth={true}
        helpText=""
        id="phoneNumber"
        name=""
        onChange={[Function]}
        placeholder="Phone Number"
        type="text"
        value="1111111111"
      />
      <Input
        autoComplete="city"
        error={true}
        fullWidth={true}
        helpText=""
        id="city"
        name=""
        onChange={[Function]}
        placeholder="City"
        type="text"
        value="Ames"
      />
      <Input
        autoComplete="state"
        error={true}
        fullWidth={true}
        helpText=""
        id="state"
        name=""
        onChange={[Function]}
        placeholder="State"
        type="text"
        value="IA"
      />
      <Input
        autoComplete="password"
        error={true}
        fullWidth={true}
        helpText=""
        id="password"
        name=""
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <Input
        error={true}
        fullWidth={true}
        helpText=""
        id="repeatPassword"
        name=""
        onChange={[Function]}
        placeholder="Repeat Password"
        type="password"
        value=""
      />
    </React.Fragment>
  }
  handleSubmit={[Function]}
  icon={<Memo(AccountCircleIcon) />}
  id="profile"
  title={
    <div
      style={
        Object {
          "display": "inline",
        }
      }
    >
      <div
        style={
          Object {
            "float": "left",
            "marginRight": "25px",
            "marginTop": "5px",
          }
        }
      >
        Your profile
      </div>
      <div
        style={
          Object {
            "float": "right",
            "marginLeft": "25px",
          }
        }
      >
        <ButtonSecondary
          disabled={false}
          fullWidth={true}
          id="signOut"
          onClick={[Function]}
          size="medium"
          type="button"
          variant="outlined"
        >
          Sign out
        </ButtonSecondary>
      </div>
    </div>
  }
/>
`;
